parla/parray/memory.py

```diff
@@ -362,8 +362,24 @@ def get_by_global_slices(self, device_id: int, global_slices: SlicesType):
             global_slices: slice/ints/tuple/list<int>, use the same format as advance indexing of numpy
 
         Return
-            :class:`cupy.ndarray` or :class:`numpy.array` object
-        """
+            :class:`cupy.ndarray` or :class:`numpy.array` object or `None` if there is no copy at that device
+        """
+        # check if a copy exists at this device
+        #
+        # This is needed for the usage of @spawn() annotation
+        # sometimes device has no copy but still need a parray slices view object in advance
+        # Example:
+        # A = parray(numpy_array)
+        # @spawn(inout=[A], placement=gpu(0))
+        # def task1():
+        #   ... a task move A to GPU. free CPU copy ...
+        # @spawn(inout=[A[0]], placement=gpu(1))
+        # def task2():
+        #   error: cpu's copy is freed -> `inout=[A[0]]` trigger an exception,
+        #   since it try a slice A at CPU (spawn() itself is happened at outer CPU task)
+        if self._buffer[device_id] is None:
+            return None
```

Comment:
or is it just returning None? does it mean that the dependency related to A[0] was removed?? In the above example, what task is the task2 dependent?

Reply:
This is a helper function. And when it return None, the caller will process the result and know this device has no copy at all, so it could generate a view object without accessing the buffer. See parray.__get_item__